File: leta/lsp/capabilities.py
from typing import Any


def get_client_capabilities() -> dict[str, Any]:
    return {
        "experimental": {
            "serverStatusNotification": True,
        },
        "workspace": {
            "applyEdit": True,
            "workspaceEdit": {
                "documentChanges": True,
                "resourceOperations": ["create", "rename", "delete"],
            },
            "symbol": {
                "dynamicRegistration": False,
                "symbolKind": {"valueSet": list(range(1, 27))},
            },
            "executeCommand": {"dynamicRegistration": False},
            "fileOperations": {
                "dynamicRegistration": False,
                "willRename": True,
                "didRename": True,
            },
        },
        "textDocument": {
            "synchronization": {
                "dynamicRegistration": False,
                "didSave": True,
            },
            "hover": {
                "dynamicRegistration": False,
                "contentFormat": ["markdown", "plaintext"],
            },
            "declaration": {"dynamicRegistration": False, "linkSupport": True},
            "definition": {"dynamicRegistration": False, "linkSupport": True},
            "typeDefinition": {"dynamicRegistration": False, "linkSupport": True},
            "implementation": {"dynamicRegistration": False, "linkSupport": True},
            "references": {"dynamicRegistration": False},
            "documentSymbol": {
                "dynamicRegistration": False,
                "symbolKind": {"valueSet": list(range(1, 27))},
                "hierarchicalDocumentSymbolSupport": True,
            },
            "codeAction": {
                "dynamicRegistration": False,
                "codeActionLiteralSupport": {
                    "codeActionKind": {
                        "valueSet": [
                            "",
                            "quickfix",
                            "refactor",
                            "refactor.extract",
                            "refactor.inline",
                            "refactor.rewrite",
                            "source",
                            "source.organizeImports",
                            "source.fixAll",
                        ]
                    }
                },
                "isPreferredSupport": True,
                "resolveSupport": {"properties": ["edit"]},
            },
            "formatting": {"dynamicRegistration": False},
            "rangeFormatting": {"dynamicRegistration": False},
            "rename": {"dynamicRegistration": False, "prepareSupport": True},
            "publishDiagnostics": {
                "relatedInformation": True,
            },
            "callHierarchy": {
                "dynamicRegistration": False,
            },
            "typeHierarchy": {
                "dynamicRegistration": False,
            },
        },
    }

File: crates/leta-lsp/src/capabilities.rs
use serde_json::{json, Value};

pub fn get_client_capabilities() -> Value {
    json!({
        "workspace": {
            "workspaceEdit": {
                "documentChanges": true,
                "resourceOperations": ["create", "rename", "delete"]
            },
            "fileOperations": {
                "willRename": true
            },
            "workspaceFolders": true,
            "configuration": true
        },
        "textDocument": {
            "definition": {
                "linkSupport": true
            },
            "declaration": {
                "linkSupport": true
            },
            "references": {
                "dynamicRegistration": false
            },
            "documentSymbol": {
                "hierarchicalDocumentSymbolSupport": true,
                "symbolKind": {
                    "valueSet": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]
                }
            },
            "rename": {
                "prepareSupport": true,
                "dynamicRegistration": false
            },
            "hover": {
                "contentFormat": ["markdown", "plaintext"]
            },
            "callHierarchy": {
                "dynamicRegistration": false
            },
            "typeHierarchy": {
                "dynamicRegistration": false
            },
            "implementation": {
                "linkSupport": true
            },
            "typeDefinition": {
                "linkSupport": true
            },
            "synchronization": {
                "didSave": true,
                "willSave": false,
                "willSaveWaitUntil": false
            },
            "completion": {
                "completionItem": {
                    "snippetSupport": false
                }
            }
        },
        "window": {
            "workDoneProgress": true
        },
        "general": {
            "positionEncodings": ["utf-32", "utf-16"]
        }
    })
}
